### ISSUE TITLE
Lasim.exe process sometimes doesn't exit

### ISSUE THREAD

**Issue by Tipoff4317** [CMahaff/lasim#39]
**Describe the Bug**
Lasim.exe process sometimes doesn't exit.  Unfortunately, I can't find a common path that is completely reproducible, but would you keep an eye on it?

I have looked at the logs both when the process exits, or when it doesn't exit.  I originally thought it only occurred when there was an error (on uploads), because it seemed to happen more, but later observations also show that even when there is no error, sometimes the process doesn't exit.

The steps that sometimes produce it is:
1) Upload
2) Click the X button on uppe right of the windows.
3) Observe that the windows disappears, but the process doesn't exit.

I have resorted to running a separate watcher process to close lasim.exe when it doesn't exit.

**Information**
LASIM Version: 0.2.2
LASIM Profile Version: 2
Lemmy Source Instance URL: https://lemm.ee/
Lemmy Source Instance Version: 0.18.4
Lemmy Target Instance URL: https://monyet.cc/
Lemmy Target Instance Version: 0.18.4

I have 4 blocked users, 5 blocked communities, and 274 subscribed communities.

**Output of LASIM**

**Comment by CMahaff** [comment]
Hmm yeah I tried it a handful of times and so far haven't seen it. Is this on Windows 10 or 11?

Also, for clarification, is this when you try to close it before the upload finishes? Or some time after the upload finishes?

**Comment by Tipoff4317** [comment]
It's a Windows 11 Pro.  

The upload is fairly quick.  I usually at least count the second that it says it takes to upload, and have to look for the process in the Process Explorer.  And then I close the Windows and watch the process.

**Comment by CMahaff** [comment]
But you are closing it after it says "upload finished", yes? Just making sure.

In any case, unfortunately don't have a Windows 11 PC at the moment so it might be hard to track down.

I'll maybe see if there is a Lemmy community that could do code review - I'm still new to Rust so it wouldn't surprise me if I'm doing `async` wrong or something - maybe it's wrong everywhere but only materializes on Windows 11 or something.

Sorry I don't have a better solution right now, but I'll keep this open until I figure it out.

**Comment by Tipoff4317** [comment]
Well, you are right.   I usually don't wait all that long because the estimated uploaded time appears to be shorter (1-2 secs) than realities (at least in tens of seconds).  Maybe this is my problem (the process doesn't exit until it finishes uploading?).  I'll let you know if I find a process that doesn't exit after it says it's finished.  

Thanks for the response.

**Comment by CMahaff** [comment]
Yeah the estimates are a bit rough because it's based on how fast the instance says it can accept connections, not how fast it does haha.

But good! Let me know what you see.

There's still something going at least slightly wrong because ideally it should stop uploading early if you close it early.  And certainly if it does run forever if you close it early that's a bug too. But I can see why those two things would be happening at least.

**Comment by Tipoff4317** [comment]
The good news is that for all the uploads that I waited to see "Finished", lasim pretty much exited immediately.  I haven't tested any that I closed the window before it said finished, though.

**Comment by CMahaff** [comment]
Fantastic! I'll leave the bug open because it should be stopping if you close the application, but I'm glad it's not an issue under normal usage.

**Comment by Tipoff4317** [comment]
Also, it seems that when I close the window before "finished", the app ultimately closes; I guess when it finishes.  The longest time it has taken so far is about 90 seconds.  

You are right; it's sort of freaky if the app doesn't shut down after closing the windows.  I would change this "bug" to an enhancement request of:

if window closed before finished => confirm shutting down (before finish) => hard exit

OR

if window closed before finished => say that the process won't exit until finish => close window => normal exit

I personally would prefer hard exit.  Leaving the app hanging around regardless seems like a possible security issue.  Imagine you ever turning this project over and the new person doing nefarious things with a background process (horror!), in Windows with no permission control / app isolation pretty much whatsoever!

Thanks.
